src/EditorFeatures/Core/Diagnostics/DiagnosticsClassificationTaggerProvider.cs

```diff
@@ -62,18 +62,47 @@ protected internal override bool IsEnabled
             => !_editorOptionsService.Factory.GlobalOptions.GetOptionValue(DefaultTextViewHostOptions.IsInContrastModeId);
 
         protected internal override bool IncludeDiagnostic(DiagnosticData data)
-            => data.CustomTags.Contains(WellKnownDiagnosticTags.Unnecessary);
+        {
+            if (!data.CustomTags.Contains(WellKnownDiagnosticTags.Unnecessary))
+            {
+                // All unnecessary code diagnostics should have the 'Unnecessary' custom tag.
+                // Below assert ensures that we do no report unnecessary code diagnostics that
+                // want to fade out multiple locations which are encoded as
+                // additional location indices in the diagnostic's property bag
+                // without the 'Unnecessary' custom tag. 
+                Debug.Assert(!TryGetUnnecessaryLocationIndices(data, out _));
+
+                return false;
+            }
+
+            // Do not fade if user has disabled the fading option corresponding to this diagnostic.
+            if (IDEDiagnosticIdToOptionMappingHelper.TryGetMappedFadingOption(data.Id, out var fadingOption))
```

Comment:
@mavasani should we also be checking this option in the LSP handler as well (since it does not go through the tagger here)?

Currently we just report the unnecessary tag to the client, and the client handles fading it - https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs,336
Should we check the option on the server side and not return the unnecessary tag, or should we ask the client to lookup the option before fading in its tagger?

Reply:
Good point. I am wondering if is it possible in the LSP handler to add unnecessary tags for additional locations which are encoded in the property bag and not tag the primary location? https://github.com/dotnet/roslyn/blob/b0de0c8e00ebf85db3c3884f2d81dfc3cb2d5a9d/src/EditorFeatures/Core/Diagnostics/DiagnosticsClassificationTaggerProvider.cs#L72-L84

If not, we need to identify how to handle these.

> Should we check the option on the server side and not return the unnecessary tag, or should we ask the client to lookup the option before fading in its tagger? If the latter we might need to migrate the setting to the editor somehow

I think it might be easier to do the former.

I have added a tracking item in https://github.com/dotnet/roslyn/issues/62666 for this work. Let's discuss offline how to tackle this.